daemon/src/service/service.rs

```diff
@@ -0,0 +1,226 @@
+
+        for object in objects {
+            for collection in &mut self.collections {
+                if collection.path().as_str() == object.as_str() {
+                    if !collection.locked() {
+                        collection.set_locked(true).await;
+                        locked.push(object.clone());
+                    }
+                }
+            }
+        }
+
+        if locked.len() == 0 {
+            locked.push(ObjectPath::default().into());
+        }
+
+        let prompt = ObjectPath::default().into(); // temporarily
+
+        Ok((locked, prompt))
+    }
+
+    pub async fn get_secrets(&self, items: Vec<OwnedObjectPath>) {
+        // WIP: not complete
+        if KEYRING.get().unwrap().n_items().await == 0 {
+            return;
+        }
+
+        // call item iface get_secret multiple times ?
+
+        for in_collection in &self.collections {
+            for item in &items {
+                if in_collection.path().as_str() == item.as_str() {}
+            }
+        }
+    }
+
+    pub async fn read_alias(&self, name: &str) -> ObjectPath {
+        let mut ret = ObjectPath::default();
+        for collection in &self.collections {
+            if collection.label() == name {
+                ret = collection.path().into();
+            }
+        }
+
+        ret
+    }
+
+    pub async fn set_alias(&mut self, name: String, collection: OwnedObjectPath) {
+        // WIP: not complete:: handle alias
+        for in_collection in &mut self.collections {
+            if in_collection.path().as_str() == collection.as_str() {
+                in_collection.set_label(name.clone()).await;
+            }
+        }
+    }
+
+    #[zbus(property, name = "Collections")]
+    pub fn collections(&self) -> Vec<ObjectPath> {
+        self.collections
+            .iter()
+            .map(|collection| collection.path())
+            .collect()
+    }
+
+    #[zbus(signal)]
+    pub async fn collection_created(ctxt: &SignalContext<'_>) -> Result<(), Error>;
+
+    #[zbus(signal)]
+    pub async fn collection_deleted(ctxt: &SignalContext<'_>) -> Result<(), Error>;
+
+    #[zbus(signal)]
+    pub async fn collection_changed(ctxt: &SignalContext<'_>) -> Result<(), Error>;
+}
+
+impl Service {
+    pub async fn new() -> Self {
+        Self {
+            collections: Vec::new(),
+            keyring: Arc::new(Keyring::load_default().await.unwrap()),
+        }
+    }
+
+    pub fn keyring(&self) -> &Keyring {
+        &self.keyring
+    }
+
+    pub async fn run(self) -> Result<(), Error> {
+        match connection::Builder::session()
+            .expect("Failed connecting to session bus")
+            .name("org.freedesktop.secrets_")
+            .expect("Failed to request name")
+            .serve_at(SECRET_SERVICE_OBJECTPATH, self)
+            .expect("Failed to serve the interface")
+            .build()
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
```

Comment:
```suggestion
        zbus::connection::Builder::session()?
            .name("org.freedesktop.secrets_")?
            .serve_at(SECRET_SERVICE_OBJECTPATH, self)?
            .build()
            .await.map_err(From::from)
    }
```
this should be better

Reply:
thanks :)

Reply:
`.await.map_err(From::from)` is introducing some wired errors,

```
.....
error[E0790]: cannot call associated function on trait without specifying the corresponding `impl` type
   --> daemon/src/service/service.rs:217:28
    |
217 |             .await.map_err(From::from)
    |                            ^^^^^^^^^^ cannot call associated function of trait

....
```

I'll get back to this later.

updated run(),

```
    pub async fn run(self) -> Result<(), Error> {
        match connection::Builder::session()?
            .name("org.freedesktop.secrets_")?
            .serve_at(SECRET_SERVICE_OBJECTPATH, self)?
            .build()
            .await // .await.map_err(From::from)
        {
            Ok(_) => Ok(()),
            Err(e) => Err(e),
        }
    }   
```